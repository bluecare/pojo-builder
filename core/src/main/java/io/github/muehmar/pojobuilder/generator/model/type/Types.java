package io.github.muehmar.pojobuilder.generator.model.type;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.PackageName;
import java.util.Optional;

public class Types {
  private Types() {}

  public static Type primitiveDouble() {
    return new Type(PrimitiveType.DOUBLE);
  }

  public static Type primitiveInt() {
    return new Type(PrimitiveType.INT);
  }

  public static Type primitiveLong() {
    return new Type(PrimitiveType.LONG);
  }

  public static Type primitiveChar() {
    return new Type(PrimitiveType.CHAR);
  }

  public static Type primitiveShort() {
    return new Type(PrimitiveType.SHORT);
  }

  public static Type primitiveByte() {
    return new Type(PrimitiveType.BYTE);
  }

  public static Type primitiveBoolean() {
    return new Type(PrimitiveType.BOOLEAN);
  }

  public static Type primitiveFloat() {
    return new Type(PrimitiveType.FLOAT);
  }

  public static Type voidType() {
    return new Type(
        DeclaredType.fromNameAndPackage(Classname.fromString("Void"), PackageName.javaLang()));
  }

  public static Type string() {
    return declaredType(Classname.fromString("String"), PackageName.javaLang());
  }

  public static Type integer() {
    return declaredType(Classname.fromString("Integer"), PackageName.javaLang());
  }

  public static Type booleanClass() {
    return declaredType(Classname.fromString("Boolean"), PackageName.javaLang());
  }

  public static Type optional(Type value) {
    return new Type(DeclaredType.optional(value));
  }

  public static Type map(Type key, Type value) {
    return new Type(
        DeclaredType.of(Classname.fromString("Map"), PackageName.javaUtil(), PList.of(key, value)));
  }

  public static Type list(Type value) {
    return new Type(
        DeclaredType.of(Classname.fromString("List"), PackageName.javaUtil(), PList.single(value)));
  }

  public static Type array(Type itemType) {
    return new Type(ArrayType.fromItemType(itemType));
  }

  public static Type varargs(Type itemType) {
    return new Type(ArrayType.varargs(itemType));
  }

  public static Type comparable(Type objType) {
    return new Type(
        DeclaredType.of(Classname.fromString("Comparable"), PackageName.javaLang(), objType));
  }

  public static Type typeVariable(Name name) {
    return new Type(TypeVariableType.ofName(name));
  }

  public static Type declaredType(Classname name, PackageName pkg) {
    return new Type(DeclaredType.fromNameAndPackage(name, pkg));
  }

  public static Type declaredType(
      Classname name, Optional<PackageName> pkg, PList<Type> typeParameters) {
    return new Type(DeclaredType.of(name, pkg, typeParameters));
  }
}
