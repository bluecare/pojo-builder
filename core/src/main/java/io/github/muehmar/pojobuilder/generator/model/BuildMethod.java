package io.github.muehmar.pojobuilder.generator.model;

import io.github.muehmar.pojobuilder.generator.model.type.Type;
import lombok.Value;

@Value
public class BuildMethod {
  Name name;
  Type returnType;
}
