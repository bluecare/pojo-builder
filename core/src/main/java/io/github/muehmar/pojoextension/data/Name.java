package io.github.muehmar.pojoextension.data;

import java.util.Objects;
import java.util.function.UnaryOperator;

/** Name of a pojo or a pojo field. */
public class Name {
  private final String value;

  private Name(String val) {
    this.value = val;
  }

  public static Name fromString(String val) {
    if (val == null || val.trim().isEmpty()) {
      throw new IllegalArgumentException("A name must no be null or empty");
    }

    return new Name(val);
  }

  public static Name concat(Name n1, Name n2) {
    return n1.append(n2);
  }

  public String asString() {
    return value;
  }

  public Name map(UnaryOperator<String> f) {
    return Name.fromString(f.apply(value));
  }

  public Name append(String append) {
    return new Name(value + append);
  }

  public Name append(Name append) {
    return new Name(value + append.value);
  }

  public Name prefix(String prefix) {
    return new Name(prefix + value);
  }

  public boolean equalsIgnoreCase(Name other) {
    return value.equalsIgnoreCase(other.value);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Name name = (Name) o;
    return Objects.equals(value, name.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value);
  }

  @Override
  public String toString() {
    return "Name{" + "value='" + value + '\'' + '}';
  }
}
