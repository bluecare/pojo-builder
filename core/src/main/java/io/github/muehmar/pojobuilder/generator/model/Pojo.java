package io.github.muehmar.pojobuilder.generator.model;

import static io.github.muehmar.pojobuilder.Booleans.not;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojobuilder.Strings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import io.github.muehmar.pojobuilder.exception.PojoBuilderException;
import io.github.muehmar.pojobuilder.generator.model.type.Type;
import java.util.Optional;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class Pojo {
  private static final PList<Name> LETTERS_AZ =
      PList.range(65, 91).map(n -> Character.toString((char) n.intValue())).map(Name::fromString);

  PojoName pojoName;
  PackageName pkg;
  PList<PojoField> fields;
  PList<Constructor> constructors;
  PList<Generic> generics;
  PList<FieldBuilder> fieldBuilders;
  Optional<BuildMethod> buildMethod;

  public PackageName getPackage() {
    return pkg;
  }

  public PList<PojoAndField> getPojoAndFields() {
    return fields.map(f -> new PojoAndField(this, f));
  }

  public PList<Name> getGenericImports() {
    return generics.flatMap(Generic::getUpperBounds).flatMap(Type::getImports);
  }

  public String getDiamond() {
    return generics.nonEmpty() ? "<>" : "";
  }

  public PList<String> getGenericTypeDeclarations() {
    return generics.map(Generic::getTypeDeclaration).map(Name::asString);
  }

  public String getGenericTypeDeclarationSection() {
    return Strings.surroundIfNotEmpty("<", getGenericTypeDeclarations().mkString(", "), ">");
  }

  public String getTypeVariablesSection() {
    return Strings.surroundIfNotEmpty(
        "<", generics.map(Generic::getTypeVariable).mkString(", "), ">");
  }

  public Name getNameWithTypeVariables() {
    return pojoName.getName().append(getTypeVariablesSection());
  }

  public String getTypeVariablesWildcardSection() {
    return Strings.surroundIfNotEmpty("<", generics.map(ignore -> "?").mkString(", "), ">");
  }

  public Optional<MatchingConstructor> findMatchingConstructor() {
    return constructors
        .flatMapOptional(c -> c.matchFields(fields).map(f -> new MatchingConstructor(c, f)))
        .headOption();
  }

  public MatchingConstructor getMatchingConstructorOrThrow() {
    return findMatchingConstructor()
        .orElseThrow(() -> new PojoBuilderException(noMatchingConstructorMessage()));
  }

  private String noMatchingConstructorMessage() {
    return String.format(
        "No matching constructor found for class/record %s."
            + " A constructor should have all the fields as arguments in the order of declaration and matching type,"
            + " where the actual type of a non-required field can be wrapped into an java.util.Optional. Furthermore"
            + "it should be accessible from within the same package, i.e. at least package-private. If a field is"
            + "instantiated in the constructor and not part of the arguments, you can annotate it with @Ignore.",
        getPojoName());
  }

  public Name findUnusedTypeVariableName() {
    final PList<Name> typeVariableNames = generics.map(Generic::getTypeVariable);
    return LETTERS_AZ
        .filter(n -> not(typeVariableNames.exists(n::equals)))
        .headOption()
        .orElseThrow(
            () ->
                new IllegalStateException(
                    "All single-letter type variables already used for generic class "
                        + getPojoName()
                        + "! If this is really a use case and should be supported, please contact the maintainer."));
  }

  public Name findUnusedTypeVariableName(Name preferred) {
    final PList<Name> typeVariableNames = generics.map(Generic::getTypeVariable);
    return LETTERS_AZ
        .cons(preferred)
        .filter(n -> not(typeVariableNames.exists(n::equals)))
        .headOption()
        .orElseThrow(
            () ->
                new IllegalStateException(
                    "All single-letter type variables already used for generic class "
                        + getPojoName()
                        + "! If this is really a use case and should be supported, please contact the maintainer."));
  }
}
