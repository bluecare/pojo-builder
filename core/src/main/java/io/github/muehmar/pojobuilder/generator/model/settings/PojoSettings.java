package io.github.muehmar.pojobuilder.generator.model.settings;

import static java.util.Optional.empty;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojobuilder.annotations.FullBuilderFieldOrder;
import io.github.muehmar.pojobuilder.annotations.OptionalDetection;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import io.github.muehmar.pojobuilder.generator.model.ClassAccessLevelModifier;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import java.util.Optional;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings {
  private static final Name CLASS_NAME_PLACEHOLDER = Name.fromString("{CLASSNAME}");
  public static final Name BUILDER_CLASS_POSTFIX = Name.fromString("Builder");
  PList<OptionalDetection> optionalDetections;
  Optional<Name> builderName;
  Optional<Name> builderSetMethodPrefix;
  ClassAccessLevelModifier builderAccessLevel;
  boolean standardBuilderEnabled;
  boolean fullBuilderEnabled;
  FullBuilderFieldOrder fullBuilderFieldOrder;

  public static PojoSettings defaultSettings() {
    return PojoSettingsBuilder.create()
        .optionalDetections(
            PList.of(OptionalDetection.OPTIONAL_CLASS, OptionalDetection.NULLABLE_ANNOTATION))
        .builderAccessLevel(ClassAccessLevelModifier.PUBLIC)
        .standardBuilderEnabled(true)
        .fullBuilderEnabled(true)
        .fullBuilderFieldOrder(FullBuilderFieldOrder.REQUIRED_FIELDS_FIRST)
        .andAllOptionals()
        .builderName(Optional.of(CLASS_NAME_PLACEHOLDER.append(BUILDER_CLASS_POSTFIX)))
        .builderSetMethodPrefix(empty())
        .build();
  }

  public Name qualifiedBuilderName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(builderName(pojo));
  }

  public Name builderName(Pojo pojo) {
    return getNameOrAppend(builderName, BUILDER_CLASS_POSTFIX, pojo);
  }

  private Name getNameOrAppend(Optional<Name> name, Name postfix, Pojo pojo) {
    return name.map(n -> n.replace(CLASS_NAME_PLACEHOLDER, getClassName(pojo)))
        .orElseGet(() -> getClassName(pojo).append(postfix));
  }

  private Name getClassName(Pojo pojo) {
    return pojo.getName().map(n -> n.replace(".", ""));
  }

  public PojoSettings withBuilderNameOpt(Name builderName) {
    return this.withBuilderName(Optional.of(builderName));
  }

  public PojoSettings withBuilderSetMethodPrefixOpt(Name builderSetMethodPrefix) {
    return this.withBuilderSetMethodPrefix(Optional.of(builderSetMethodPrefix));
  }

  public PojoSettings overrideOptionalDetection(
      Optional<PList<OptionalDetection>> optionalDetections) {
    return optionalDetections.map(this::withOptionalDetections).orElse(this);
  }

  public PojoSettings overrideBuilderName(Optional<Name> builderName) {
    return builderName.map(this::withBuilderNameOpt).orElse(this);
  }

  public PojoSettings overrideBuilderSetMethodPrefix(Optional<Name> builderSetMethodPrefix) {
    return builderSetMethodPrefix
        .map(ignore -> this.withBuilderSetMethodPrefix(builderSetMethodPrefix))
        .orElse(this);
  }

  public PojoSettings overrideBuilderAccessLevel(
      Optional<ClassAccessLevelModifier> classAccessLevelModifier) {
    return classAccessLevelModifier.map(this::withBuilderAccessLevel).orElse(this);
  }
}
