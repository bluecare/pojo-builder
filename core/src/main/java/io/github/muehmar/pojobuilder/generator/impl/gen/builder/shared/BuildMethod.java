package io.github.muehmar.pojobuilder.generator.impl.gen.builder.shared;

import static io.github.muehmar.codegenerator.Generator.constant;
import static io.github.muehmar.codegenerator.java.JavaModifier.PUBLIC;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaGenerators;
import io.github.muehmar.pojobuilder.generator.impl.gen.ThrowsGenerator;
import io.github.muehmar.pojobuilder.generator.model.Constructor;
import io.github.muehmar.pojobuilder.generator.model.FactoryMethod;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.matching.MatchingConstructor;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.generator.model.type.QualifiedClassname;
import io.github.muehmar.pojobuilder.generator.model.type.Type;
import java.util.function.Function;

public class BuildMethod {
  private BuildMethod() {}

  public static Generator<Pojo, PojoSettings> standardBuilderBuildMethod() {
    return buildMethod(constant("return builder.build();"));
  }

  public static Generator<Pojo, PojoSettings> buildMethod(
      Generator<Pojo, PojoSettings> buildMethodContent) {
    final Function<Pojo, Object> createReturnType =
        p ->
            p.getBuildMethod()
                .map(io.github.muehmar.pojobuilder.generator.model.BuildMethod::getReturnType)
                .map(Type::getTypeDeclaration)
                .map(Name::asString)
                .orElse(
                    String.format(
                        "%s%s",
                        p.getPkg().equals(p.getPojoClassname().getPkg())
                            ? ""
                            : p.getPojoClassname().getPkg() + ".",
                        p.getPojoNameWithTypeVariables()));
    return JavaGenerators.<Pojo, PojoSettings>methodGen()
        .modifiers(PUBLIC)
        .noGenericTypes()
        .returnType(createReturnType)
        .methodName("build")
        .noArguments()
        .throwsExceptions(throwsGenerator())
        .content(buildMethodContent)
        .build();
  }

  private static Generator<Pojo, PojoSettings> throwsGenerator() {
    return Generator.<Pojo, PojoSettings>emptyGen()
        .append(ThrowsGenerator.throwsGenerator(), BuildMethod::getAllBuildExceptionsFromPojo);
  }

  private static PList<QualifiedClassname> getAllBuildExceptionsFromPojo(
      Pojo pojo, PojoSettings settings) {
    return getExceptionsFromFactoryMethod(pojo)
        .concat(getConstructorExceptionsFromPojo(pojo, settings))
        .concat(getBuildMethodExceptionsFromPojo(pojo))
        .distinct(Function.identity());
  }

  private static PList<QualifiedClassname> getExceptionsFromFactoryMethod(Pojo pojo) {
    return pojo.getFactoryMethod().map(FactoryMethod::getExceptions).orElseGet(PList::empty);
  }

  private static PList<QualifiedClassname> getConstructorExceptionsFromPojo(
      Pojo pojo, PojoSettings settings) {
    if (pojo.getFactoryMethod().isPresent()) {
      return PList.empty();
    }
    return pojo.findMatchingConstructor(settings.getFieldMatching())
        .getFirstMatchingConstructor()
        .map(MatchingConstructor::getConstructor)
        .map(Constructor::getExceptions)
        .orElseGet(PList::empty);
  }

  private static PList<QualifiedClassname> getBuildMethodExceptionsFromPojo(Pojo pojo) {
    return pojo.getBuildMethod()
        .map(io.github.muehmar.pojobuilder.generator.model.BuildMethod::getExceptions)
        .orElseGet(PList::empty);
  }
}
