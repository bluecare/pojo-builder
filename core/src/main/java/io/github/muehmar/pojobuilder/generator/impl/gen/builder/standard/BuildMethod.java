package io.github.muehmar.pojobuilder.generator.impl.gen.builder.standard;

import static io.github.muehmar.codegenerator.java.JavaModifier.PUBLIC;

import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaGenerators;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.generator.model.type.Type;
import java.util.function.Function;

class BuildMethod {
  private BuildMethod() {}

  public static Generator<Pojo, PojoSettings> buildMethod() {
    final Function<Pojo, String> createReturnType =
        p ->
            p.getBuildMethod()
                .map(io.github.muehmar.pojobuilder.generator.model.BuildMethod::getReturnType)
                .map(Type::getTypeDeclaration)
                .orElseGet(p::getNameWithTypeVariables)
                .asString();
    return JavaGenerators.<Pojo, PojoSettings>methodGen()
        .modifiers(PUBLIC)
        .noGenericTypes()
        .returnType(createReturnType)
        .methodName("build")
        .noArguments()
        .content("return builder.build();")
        .build();
  }
}
