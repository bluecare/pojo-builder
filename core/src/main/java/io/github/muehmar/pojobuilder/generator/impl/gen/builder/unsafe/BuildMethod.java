package io.github.muehmar.pojobuilder.generator.impl.gen.builder.unsafe;

import static io.github.muehmar.codegenerator.java.JavaModifier.PUBLIC;
import static io.github.muehmar.pojobuilder.Booleans.not;
import static io.github.muehmar.pojobuilder.generator.impl.gen.instantiation.ConstructorCallGenerator.constructorCallGenerator;

import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaGenerators;
import io.github.muehmar.pojobuilder.generator.impl.gen.instantiation.FactoryMethodCallGenerator;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.generator.model.type.Type;
import java.util.function.Function;

class BuildMethod {
  private BuildMethod() {}

  public static Generator<Pojo, PojoSettings> buildMethod() {
    final Function<Pojo, Object> createReturnType =
        p ->
            p.getBuildMethod()
                .map(io.github.muehmar.pojobuilder.generator.model.BuildMethod::getReturnType)
                .map(Type::getTypeDeclaration)
                .map(Name::asString)
                .orElse(
                    String.format(
                        "%s%s",
                        p.getPkg().equals(p.getPojoClassname().getPkg())
                            ? ""
                            : p.getPojoClassname().getPkg() + ".",
                        p.getPojoNameWithTypeVariables()));
    return JavaGenerators.<Pojo, PojoSettings>methodGen()
        .modifiers(PUBLIC)
        .noGenericTypes()
        .returnType(createReturnType)
        .methodName("build")
        .noArguments()
        .doesNotThrow()
        .content(buildMethodContent())
        .build();
  }

  private static Generator<Pojo, PojoSettings> buildMethodContent() {
    return constructorCall().append(factoryMethodCall());
  }

  private static Generator<Pojo, PojoSettings> factoryMethodCall() {
    return Generator.<Pojo, PojoSettings>emptyGen()
        .append((p, s, w) -> w.print("return "))
        .append(FactoryMethodCallGenerator.factoryMethodCallGenerator())
        .filter(pojo -> pojo.getFactoryMethod().isPresent());
  }

  private static Generator<Pojo, PojoSettings> constructorCall() {
    final Generator<Pojo, PojoSettings> returnGenerator =
        (p, s, w) ->
            p.getBuildMethod()
                .map(
                    bm ->
                        w.println(
                            "return %s.%s(instance);",
                            p.getPojoClassname().getName(), bm.getName()))
                .orElse(w.println("return instance;"));
    return Generator.<Pojo, PojoSettings>emptyGen()
        .append((p, s, w) -> w.println("final %s instance =", p.getPojoNameWithTypeVariables()))
        .append(constructorCallGenerator(), 2)
        .append(returnGenerator)
        .filter(pojo -> not(pojo.getFactoryMethod().isPresent()));
  }
}
