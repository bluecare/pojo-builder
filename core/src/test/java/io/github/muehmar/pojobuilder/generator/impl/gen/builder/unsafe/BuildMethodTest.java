package io.github.muehmar.pojobuilder.generator.impl.gen.builder.unsafe;

import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static io.github.muehmar.pojobuilder.generator.impl.gen.builder.unsafe.BuildMethod.buildMethod;
import static io.github.muehmar.pojobuilder.snapshottesting.SnapshotUtil.writerSnapshot;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.PList;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.generator.Pojos;
import io.github.muehmar.pojobuilder.generator.model.BuildMethod;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.generator.model.type.Types;
import io.github.muehmar.pojobuilder.snapshottesting.SnapshotTest;
import java.util.Optional;
import org.junit.jupiter.api.Test;

@SnapshotTest
class BuildMethodTest {
  private Expect expect;

  @Test
  @SnapshotName("sample")
  void buildMethod_when_calledWithSample_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator = buildMethod();
    final String output =
        generator.generate(Pojos.sample(), PojoSettings.defaultSettings(), javaWriter()).asString();

    expect.toMatchSnapshot(output);
  }

  @Test
  @SnapshotName("genericSample")
  void buildMethod_when_calledWithGenericSample_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator = buildMethod();
    final String output =
        generator
            .generate(Pojos.genericSample(), PojoSettings.defaultSettings(), javaWriter())
            .asString();

    expect.toMatchSnapshot(output);
  }

  @Test
  @SnapshotName("genericSampleAndBuildMethod")
  void buildMethod_when_calledWithGenericSampleAndBuildMethod_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator = buildMethod();
    final io.github.muehmar.pojobuilder.generator.model.BuildMethod buildMethod =
        new BuildMethod(Name.fromString("customBuildMethod"), Types.string(), PList.empty());
    final Pojo pojo = Pojos.genericSample().withBuildMethod(Optional.of(buildMethod));
    final String output =
        generator.generate(pojo, PojoSettings.defaultSettings(), javaWriter()).asString();

    expect.toMatchSnapshot(output);
  }

  @Test
  @SnapshotName("sampleAndConstructorWithOptionalArgument")
  void
      buildMethod_when_generatorUsedWithSamplePojoAndConstructorWithOptionalArgument_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator = buildMethod();
    final Writer writer =
        generator.generate(
            Pojos.sampleWithConstructorWithOptionalArgument(),
            PojoSettings.defaultSettings(),
            javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("sampleWithFactoryMethod")
  void buildMethod_when_sampleWithFactoryMethod_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator = buildMethod();
    final Writer writer =
        generator.generate(
            Pojos.factoryMethodSample(), PojoSettings.defaultSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
