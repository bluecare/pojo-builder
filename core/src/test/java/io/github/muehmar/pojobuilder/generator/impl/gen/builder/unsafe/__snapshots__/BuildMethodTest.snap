genericSample=[
public Customer<T, S> build() {
  final Customer<T, S> instance =
      new Customer<T, S>(id, data, additionalData);
  return instance;
}
]


genericSampleAndBuildMethod=[
public String build() {
  final Customer<T, S> instance =
      new Customer<T, S>(id, data, additionalData);
  return Customer.customBuildMethod(instance);
}
]


sample=[
public Customer build() {
  final Customer instance =
      new Customer(id, username, nickname);
  return instance;
}
]


sampleAndConstructorWithOptionalArgument=[
java.util.Optional

public Customer build() {
  final Customer instance =
      new Customer(id, username, Optional.ofNullable(nickname));
  return instance;
}
]


sampleWithFactoryMethod=[
io.github.muehmar.exception.SomeException
java.util.Optional

public io.github.muehmar.Customer build() throws SomeException {
  return Pojos.Customer.create(id, username, Optional.ofNullable(nickname));
}
]